reset_column_information now clears database statement cache for each…

kum-deepak opened the pull request:
The reset_column_information in ActiveRecord::Base does not clear cached database statements corresponding to associations.

The issue was encountered during test cases for gem webdack-uuid_migration. This gem, for Postgres only, updates a database in-place to use UUID instead of integers/serial.

This pull request has test case to show the issue. The Gemfile includes webdack-uuid_migration gem. If that poses an issue, I will copy the relevant code from the Gem into the test case itself.

All test cases for ActiveRecord has been executed and everything succeeds without any additional warnings.

Please let me know if any additional information is needed.

kum-deepak committed:
reset_column_information now clears database statement cache for each association.

matthewd commented:
> The Gemfile includes webdack-uuid_migration gem. If that poses an issue, I will copy the relevant code from the Gem into the test case itself.

Yes, that poses an issue.

We don't want a copy of the gem, either. The test should be as small as possible: just enough to demonstrate the problem (so, just changing the type of one or two columns, as far as I can see). And the test should make a rudimentary effort to look slightly like some of the others.

kum-deepak committed:
Removed gem dependency on webdack-uuid_migration.

kum-deepak committed:
Removed gem dependency on webdack-uuid_migration.

kum-deepak commented:
I have removed the gem dependency. Please suggest if more changes are needed.

prathamesh-sonpatki commented on activerecord/test/cases/associations/belongs_to_associations_statement_cache_test.rb:
```diff
@@ -0,0 +1,99 @@
+#
+# The bug was discovered during testing of gem webdack-uuid_migration
+# That gem allows in place upgrading of Postgres databses to use UUIDs
+# This bug most likely had appeared in some version of 4.x.x
```
All of this information can go in the commit message itself rather than here :)

kum-deepak committed:
The reset_column_information does not clear some cached database statements. This affects
cases where the underlying column types changes. In this case ActiveRecord throws an
exception complaining StatementInvalid.

This testcase uses its own models and migrations as it updates the underlying column types
from integer to uuid.

Names of models and tables are quirky student -> dtudent and city -> dity
to avoid conflict with any other test cases or models

The bug was discovered during testing of gem webdack-uuid_migration
That gem allows in place upgrading of Postgres databses to use UUIDs
This bug most likely had appeared in some version of 4.x.x

kum-deepak commented:
Any updates? Are any more changes needed?

kum-deepak commented:
Any updates? Are any more changes needed?

kum-deepak commented:
Any updates? Are any more changes needed?

kum-deepak commented:
It is about 3 months since this was submitted, any updates?

Has this been rejected?

rafaelfranca commented on activerecord/test/cases/associations/belongs_to_associations_statement_cache_test.rb:
```diff
@@ -0,0 +1,81 @@
+require 'cases/helper'
+
+class Dity < ActiveRecord::Base
```
Could you try to use the current models in the Active Record test suite to reproduce this problem?

rafaelfranca commented on activerecord/test/cases/associations/belongs_to_associations_statement_cache_test.rb:
```diff
@@ -0,0 +1,81 @@
+  def change
+    reversible do |dir|
+      dir.up do
+        enable_extension 'uuid-ossp'
```
IS there any way to reproduce this issue without using uuid-ossp?
